File: shared/src/main/java/de/dreier/mytargets/shared/models/SelectableZone.java
/*
 * Copyright (C) 2016 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.models;

import android.support.annotation.NonNull;

import de.dreier.mytargets.shared.targets.zone.ZoneBase;

public class SelectableZone implements Comparable<SelectableZone> {
    public final int index;
    public final ZoneBase zone;
    public final int points;
    public final String text;

    public SelectableZone(int index, ZoneBase zone, String text, int points) {
        this.index = index;
        this.zone = zone;
        this.points = points;
        this.text = text;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) {
            return true;
        }
        if (o == null || getClass() != o.getClass()) {
            return false;
        }

        SelectableZone that = (SelectableZone) o;
        return points == that.points && text.equals(that.text);
    }

    @Override
    public int hashCode() {
        return 31 * points + text.hashCode();
    }

    @Override
    public int compareTo(@NonNull SelectableZone another) {
        if (another.index == index) {
            return 0;
        } else if (another.index >= 0 && index >= 0) {
            return index - another.index;
        } else {
            return another.index - index;
        }
    }
}

File: shared/src/main/java/de/dreier/mytargets/shared/models/SelectableZone.kt
/*
 * Copyright (C) 2017 Florian Dreier
 *
 * This file is part of MyTargets.
 *
 * MyTargets is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2
 * as published by the Free Software Foundation.
 *
 * MyTargets is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 */

package de.dreier.mytargets.shared.models

import de.dreier.mytargets.shared.targets.zone.ZoneBase

class SelectableZone(
        val index: Int,
        val zone: ZoneBase,
        val text: String,
        val points: Int
) : Comparable<SelectableZone> {

    override fun compareTo(other: SelectableZone): Int {
        return if (other.index == index) {
            0
        } else if (other.index >= 0 && index >= 0) {
            index - other.index
        } else {
            other.index - index
        }
    }

    /**
     * Used in de.dreier.mytargets.shared.models.db.End.Companion#getRoundScores
     */
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as SelectableZone

        if (text != other.text) return false
        if (points != other.points) return false

        return true
    }

    override fun hashCode(): Int {
        return 31 * text.hashCode() + points
    }
}
